### PR TITLE
Enable node-based UIs in LSP

### PR DESCRIPTION
**Pull request by sdedic**
This PR is rather a prototype, or start of a larger work. Actions etc need to be somehow implemented on the protocol. At this moment, just node appearance and structure is transmitted to a LSP client for the display. The only supported action at the moment is open - for nodes that represent files. Project View is the first view to use this new LSP service, but more node hierarchies can be defined & served.

The appearance in a LSP client (vscode in this case) could be as follows:
vscode-nodes

### CODE REVIEW

File: java/java.lsp.server/vscode/src/explorer.ts [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+      this.images.set(nodeData.iconIndex, uri);
+      return uri;
+    } else {
+      return this.images.get(nodeData.iconIndex);
+    }
+  }
+}
+
+let treeService : Promise<TreeViewService>;
+let promisedTreeService : (ts : TreeViewService) => void;
```
Comment (JaroslavTulach) [inline java/java.lsp.server/vscode/src/explorer.ts:62]: Please eliminate all static state from this `.ts` file and other `.ts` files. The only static state shall be in `explorer.ts`, currently:
```ts
let client: Promise<LanguageClient>;
let testAdapter: NbTestAdapter | undefined;
let nbProcess : ChildProcess | null = null;
let debugPort: number = -1;
let consoleLog: boolean = !!process.env['ENABLE_CONSOLE_LOG'];
```
Make the `TreeViewService` a field in `LanguageClient`, if possible.

File: ide/projectuiapi.base/src/org/netbeans/api/project/ui/OpenProjects.java [hunk 2]
```diff
@@ -315,5 +314,24 @@ public void addProjectGroupChangeListener( @NonNull ProjectGroupChangeListener l
...
     }
-    
+
+    /** Creates new {@link ExplorerManager} to show and control
+     * logical structure of a project.
+     *
+     * @return new instance of the manager
+     * @since 1.100
+     */
+    public ExplorerManager createLogicalView() {
```
Comment (JaroslavTulach) [inline ide/projectuiapi.base/src/org/netbeans/api/project/ui/OpenProjects.java:324]: This API has to be documented in apichanges. Sorry for not doing that to begin with.

File: java/java.lsp.server/nbcode/integration/src/org/netbeans/modules/nbcode/integration/layer.xml [hunk 3]
```diff
@@ -21,6 +21,11 @@
 -->
 <!DOCTYPE filesystem PUBLIC "-//NetBeans//DTD Filesystem 1.2//EN" "http://www.netbeans.org/dtds/filesystem-1_2.dtd">
 <filesystem>
+    <folder name="Explorers">
+        <folder name="_all">
+            <file name="standard-cookies.contextValues" url="standard-cookies.contextValues"/>
```
Comment (JaroslavTulach) [inline java/java.lsp.server/nbcode/integration/src/org/netbeans/modules/nbcode/integration/layer.xml:26]: Interesting API. Is it documented somewhere? I'd rather use `<attr ..../>` than content of a file, however.

File: java/java.lsp.server/nbcode/integration/src/org/netbeans/modules/nbcode/integration/standard-cookies.contextValues [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+# under the License.
+
+# Use fully qualified names, one class per line. Syntax is:
+#   qualified.name=value    : value will be used instead of qualified name.
+#   [prefix.]qualified.name : the prefix will be stripped
+#   [prefix=subst]          : the prefix in all classes will be replaced by subst, possibly empty
+#   * as the first character: watch for changes in this interface, fire changes.
+# Names will be converted to "class:<name>" tokens.
+
+org.netbeans.api.actions.Openable
```
Comment (JaroslavTulach) [inline java/java.lsp.server/nbcode/integration/src/org/netbeans/modules/nbcode/integration/standard-cookies.contextValues:25]: None of the descriptions at line 18-21 seem to match `org.netbeans.api.actions.Openable` - I have no idea what it does.

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/explorer/ProjectExplorer.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+import org.openide.explorer.ExplorerManager;
+import org.openide.util.Lookup;
+import org.openide.util.RequestProcessor;
+import org.openide.util.lookup.ServiceProvider;
+
+/**
+ *
+ * @author sdedic
+ */
+@ServiceProvider(path = "Explorers/" + ProjectExplorer.ID_PROJECT_LOGICAL_VIEW, service = ExplorerManagerFactory.class)
```
Comment (JaroslavTulach) [inline java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/explorer/ProjectExplorer.java:34]: Please document somewhere (in this PR?) steps to add new treeview even it is just a private API in `java.lsp.server` module.

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/explorer/api/ExplorerManagerFactory.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+import java.util.concurrent.CompletionStage;
+import org.openide.explorer.ExplorerManager;
+import org.openide.util.Lookup;
+
+/**
+ * Factory that produces an ExplorerManager that serves node in a specified tree structure. The
+ * Factory has to be registered as a named service on path {@code Explorers/&lt;id>}; first registration
+ * wins.
+ * <p>
+ * Note: this interface will become an API independent of LSP.
```
Comment (JaroslavTulach) [inline java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/explorer/api/ExplorerManagerFactory.java:30]: Great. Include steps & references to what needs to be done, please.

File: java/java.lsp.server/vscode/.vscode/launch.json [hunk 7]
```diff
@@ -31,7 +31,10 @@
 			"outFiles": [
 				"${workspaceFolder}/out/test/**/*.js"
 			],
-			"preLaunchTask": "${defaultBuildTask}"
+			"preLaunchTask": "${defaultBuildTask}",
+			"env": { 
+				"netbeans.extra.options" : "-J-Dproject.limitScanRoot=${workspaceFolder}/out"
```
Comment (JaroslavTulach) [inline java/java.lsp.server/vscode/.vscode/launch.json:36]: An API to document as `<api category="systemproperty" />` somewhere.

File: java/java.lsp.server/vscode/src/explorer.ts [hunk 8]
```diff
@@ -0,0 +1,227 @@
+import * as vscode from 'vscode';
```
Comment (JaroslavTulach) [inline java/java.lsp.server/vscode/src/explorer.ts:1]: File without a license header!?

**Review by JaroslavTulach — APPROVED** [review]
Thanks for your great work, Sváťo. Thanks for fighting hard to stabilize the tests!

The API changes outside of `java.lsp.server` are minimal - e.g. you have my approval. However please document them.

The APIs inside of `java.lsp.server` look fine and anyway, we'll have a time to polish them as the time goes.

**Comment by sdedic** [comment]
In the first round I've squashed the initial implementation and left some feature-unrealated changes separate.

File: java/java.lsp.server/vscode/src/explorer.ts (hunk 1 above)
Comment (sdedic) [reply java/java.lsp.server/vscode/src/explorer.ts]: Done (by you :-)) in 60a33320

File: java/java.lsp.server/nbcode/integration/src/org/netbeans/modules/nbcode/integration/layer.xml (hunk 3 above)
Comment (sdedic) [reply java/java.lsp.server/nbcode/integration/src/org/netbeans/modules/nbcode/integration/layer.xml:26]: Not documented yet, and I'd wait with publication for the next few PRs - I am going to play with DB explorer nodes, and probably provide some project-related actions. I'd expect some changes.

File: java/java.lsp.server/nbcode/integration/src/org/netbeans/modules/nbcode/integration/standard-cookies.contextValues (hunk 4 above)
Comment (sdedic) [reply java/java.lsp.server/nbcode/integration/src/org/netbeans/modules/nbcode/integration/standard-cookies.contextValues:25]: nodes that can be 'opened' somehow will manifest this using `class:Openable` at LSP client side.

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/explorer/api/ExplorerManagerFactory.java (hunk 6 above)
Comment (sdedic) [reply java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/explorer/api/ExplorerManagerFactory.java:39]: Steps described in javadoc.

File: java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/explorer/ProjectExplorer.java (hunk 5 above)
Comment (sdedic) [reply java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/explorer/ProjectExplorer.java:34]: Steps described in `ExplorerManagerFactory`'s javadoc.

File: java/java.lsp.server/vscode/src/explorer.ts (hunk 8 above)
Comment (sdedic) [reply java/java.lsp.server/vscode/src/explorer.ts:1]: Fixed headers in all `.ts` tests.

File: java/java.lsp.server/vscode/.vscode/launch.json (hunk 7 above)
Comment (sdedic) [reply java/java.lsp.server/vscode/.vscode/launch.json]: OK, documented in project api's `arch.xml`

File: java/java.lsp.server/nbcode/integration/src/org/netbeans/modules/nbcode/integration/standard-cookies.contextValues (hunk 4 above)
Comment (JaroslavTulach) [reply java/java.lsp.server/nbcode/integration/src/org/netbeans/modules/nbcode/integration/standard-cookies.contextValues:25]: OK, let's leave it as it is for now.

File: java/java.lsp.server/vscode/package.json [hunk 9]
```diff
@@ -440,6 +452,14 @@
 					"command": "graalvm.pause.script",
 					"when": "nbJavaLSReady"
 				}
+			],
+			"view/title": [
```
Comment (dbalek) [inline java/java.lsp.server/vscode/package.json:456]: Empty array?

**Comment by JaroslavTulach** [comment]
Sváťo, please update `CHANGELOG.md` and `READMEmd`.

File: java/java.lsp.server/vscode/src/protocol.ts [hunk 10]
```diff
@@ -19,11 +19,14 @@
 'use strict';
 
 import * as vscode from 'vscode';
+
 import {
     ProtocolNotificationType,
     ProtocolRequestType,
+    RequestType,
```
Comment (dbalek) [inline java/java.lsp.server/vscode/src/protocol.ts:26]: Is the import used?

**Review by dbalek — APPROVED** [review]

**Comment by sdedic** [comment]
Fails just on `SecureURLResourceRetrieverTest`, which is an independent bug - that one should be probably fixed ASAP.

**Pull request merged**